jaxrs-core/src/main/groovy/org/grails/plugins/jaxrs/servlet/http/ReusableStreamHttpServletRequestWrapper.groovy

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2009 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.grails.plugins.jaxrs.servlet.http
+
+import groovy.transform.CompileStatic
+import javax.servlet.ServletInputStream
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletRequestWrapper
+import org.apache.commons.io.IOUtils
+import org.grails.plugins.jaxrs.servlet.ReusableServletInputStream
+
+/** 
+ *
+ * @author Alex Stoia
+ */
+@CompileStatic    
+class ReusableStreamHttpServletRequestWrapper extends HttpServletRequestWrapper {
+    private BufferedInputStream wrappedStream
+    private ServletInputStream usedStream
+        
+    ReusableStreamHttpServletRequestWrapper(HttpServletRequest request) {
+        super(request)
+        ServletInputStream originalStream = request.getInputStream()
+        usedStream = new ReusableServletInputStream(originalStream)
+        usedStream.mark(Integer.MAX_VALUE)
+        // read the stream contents, otherwise this will not be available in 
+        // the getInputStream method. Also tried reading one byte, works on 
+        // first request, fails on all subsequent ones. Got no idea why at
+        // the moment
+        IOUtils.toString(usedStream, 'UTF-8')
```

Comment:
This is concerning to me because this ultimately will require all payloads to be read into memory for the duration of the request. This doesn't allow authors to efficiently pipe streams together, and can potentially cause large memory requirements, depending on the expected payload sizes.